// discord-button-page/src/index.js
let DJSVersions = require('../../../package.json').dependencies['discord.js'] || require('../../../package.json').devDependencies['discord.js'];
DJSVersions = DJSVersions.toString();

let versions = DJSVersions.split(".")[0];

if (versions === "^12"){
  module.exports = {
    MessageButtonPages: require('./structures/v12/MessageButtonPages.js'),
    MessageReactionPages: require('./structures/v12/MessageReactionPages.js'),
    author: require("../package.json").author,
    version: require("../package.json").version
  }
} else if (versions === "^13"){
  module.exports = {
    InteractionButtonPages: require('./structures/v13/InteractionButtonPages.js'),
    InteractionReactionPages: require('./structures/v13/InteractionReactionPages.js'),
    MessageButtonPages: require('./structures/v13/MessageButtonPages.js'),
    MessageReactionPages: require('./structures/v13/MessageReactionPages.js'),
    author: require("../package.json").author,
    version: require("../package.json").version
  }
}
