// discord-button-page/src/structures/v12/MessageReactionPages.js
const { MessageEmbed, Message } = require("discord.js");
const { Error, RangeError } = require("../../errors");
const Util = require("../../util/Util.js");
const ClassFunction = require("../../util/ClassFunction.js");

/**
 * Setup message command reaction pages with this class
 */
class MessageReactionPages {

  constructor(data = {}) {
    this.setup(data);
  }

  setup(data) {
    /**
     * The embed of Discord.JS
     * @type {?array<MessageEmbed>}
     */
    this.embeds = data.embeds || null;
    ClassFunction.EmbedsOptions(data.embeds);
    
    /**
     * The duration of this MessageReactionPages
     * @type {?number}
     */
    this.duration = data.duration || 30000;
    ClassFunction.DurationOptions(data.duration);
    

    /**
     * The counting of this MessageReactionPages
     * @type {?boolean}
     */
    this.countPage = data.countPage || true;
    ClassFunction.CountPagesOptions(data.countPage);

    /**
     * The emoji of this MessageReactionPages
     * @type {?array<string>}
     */
    this.emoji = {
      previous: {
        name: undefined,
        id: undefined
      },
      stop: {
        name: undefined,
        id: undefined
      },
      next: {
        name: undefined,
        id: undefined
      }
    };
    ClassFunction.EmojiOptions(data.emoji, this.emoji);

    /**
     * The currentPage of this MessageReactionPages
     * @type {?number}
     */
    this.currentPage = 0;

    return this;
  }

  /**
   * Adds embed to create page
   * @param {string} [embed] Embed option in array
   * @returns {MessageReactionPages[]}
   */
  setEmbeds(embed) {
    ClassFunction.EmbedsOptions(embed);
    
    try {
      this.embeds = embed.map(embeds => new MessageEmbed(embeds));
    } catch(e) {
      throw new Error("INVALID_EMBED");
    }
    return this;
  }

  /**
   * Adds duration in this option
   * @param {number} time The time duration
   * @returns {MessageReactionPages}
   */
  setDuration(time) {
    ClassFunction.DurationOptions(time);

    this.duration = time;
    return this;
  }

  /**
   * Adds count page of your embeds
   * @param {boolean} [count=true] Count all of your embeds
   * @returns {MessageReactionPages}
   */
  setCountPage(counting = true) {
    ClassFunction.CountPagesOptions(counting);

    if (counting === false) this.countPage = false;
    else this.countPage = true;
    return this;
  }

  /**
   * Change emoji of the button
   * @param {string} emoji Custom emoji Previous, Stop, Next
   * @returns {MessageButtonPages[]}
   */
  setEmojiReaction(emoji) {
    ClassFunction.EmojiOptions(emoji, this.emoji)
    return this;
  }



  /**
   * Build page reaction
   * @param {Message} message Syntax with message
   * @returns {MessageReactionPages}
   */
  build(message) {
    if (!message)
      throw new RangeError("REQUIRE", "Message", "message");
    
    if (message && message.type !== "DEFAULT")
      throw new RangeError("ASYNC_ERROR", "Message");

    if (this.countPage === true) this.embeds[0].setFooter(`Page 1 of ${this.embeds.length}`).setTimestamp(Date.now());
    if (this.countPage === false) this.embeds[0];

    Util.Reaction(message, this.embeds, this.reply, this.duration, this.countPage, this.currentPage, this.emoji, "CHANNEL", "MESSAGE", "v12");
  }

  /**
   * data of this class
   * @returns {MessageReactionPages}
   */
  toJSON() {
    return {
      embeds: this.embeds,
      duration: this.duration,
      countPage: this.countPage,
      emoji: this.emoji && {
        previous: {
          name: this.emoji.previous.name,
          id: this.emoji.previous.id,
        },
        stop: {
          name: this.emoji.stop.name,
          id: this.emoji.stop.id,
        },
        next: {
          name: this.emoji.next.name,
          id: this.emoji.next.id,
        },
      },
    }
  }
}

module.exports = MessageReactionPages;
